src/cocaine-app/infrastructure.py

```diff
@@ -368,61 +445,65 @@ def set_ns_settings(self, namespace, settings):
         del settings['namespace']
         self.ns_settings[namespace] = settings
 
-    def _update_group(self, group_id, new_nodes=None, new_couple=None, record_type=None):
-        group = self.state[group_id]
+    def _update_group(self,
+                      group_history,
+                      new_nodes=None,
+                      new_couple=None,
+                      record_type=None):
+
         if new_nodes is not None:
-            new_nodes_state = {'set': new_nodes,
-                               'timestamp': time.time()}
-            new_nodes_state['type'] = record_type or self.HISTORY_RECORD_AUTOMATIC
+            new_nodes.timestamp = time.time()
+            new_nodes.type = record_type or GroupStateRecord.HISTORY_RECORD_AUTOMATIC
 
-            group['nodes'].append(new_nodes_state)
+            group_history.nodes.append(new_nodes)
+            group_history._dirty = True
 
         if new_couple is not None:
-            new_couples_state = {'couple': new_couple,
-                                 'timestamp': time.time()}
-            group['couples'].append(new_couples_state)
-
-        eid = self.meta_session.transform(keys.MM_ISTRUCT_GROUP % group_id)
-        logger.info('Updating state for group %s' % group_id)
-        self.meta_session.update_indexes(eid, [keys.MM_GROUPS_IDX],
-                                              [self._serialize(group)]).get()
-
-    def detach_node(self, group_id, host, port, backend_id, record_type=None):
-        with self.__state_lock:
-            if group_id not in self.state:
-                raise ValueError(
-                    'Node backend {0}:{1}/{2} not found in group {3} history state'.format(
-                        host, port, backend_id, group_id
-                    )
+            new_couple.timestamp = time.time()
```

Comment:
Shouldn't you get the current time only once and use it for both `new_nodes.timestamp` and `new_couple.timestamp`?